From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 16 Jan 2021 22:44:20 -0600
Subject: [PATCH] Convert EntityTag attributes on creature spawn

Entity conversion does not run on the EntityTag.
This is a serious flaw in this SpawnEgg EntityTag features

diff --git a/src/main/java/net/minecraft/world/entity/EntityType.java b/src/main/java/net/minecraft/world/entity/EntityType.java
index 6b209ba75d0f9a30beb20117b581f87936097b59..2f11b037356482ec6050593e78f2d59c0825b71a 100644
--- a/src/main/java/net/minecraft/world/entity/EntityType.java
+++ b/src/main/java/net/minecraft/world/entity/EntityType.java
@@ -378,6 +378,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T> {
             }
         }
         // Paper end
+        convertNBT(nbttagcompound); // EMC
         T t0 = this.create(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1);
         // EMC start - if false the spawn was cancelled, add new event
         if ((spawnReason == CreatureSpawnEvent.SpawnReason.SPAWNER_EGG || spawnReason == CreatureSpawnEvent.SpawnReason.DISPENSE_EGG) && itemstack != null && t0 != null) {
@@ -403,6 +404,54 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T> {
         return t0;
     }
 
+    // EMC start
+    private static final java.util.Map<String, String> attributeConversion = new com.google.common.collect.ImmutableMap.Builder<String, String>()
+        .put("generic.maxHealth", "generic.max_health")
+        .put("horse.jumpStrength", "horse.jump_strength")
+        .put("generic.followRange", "generic.follow_range")
+        .put("generic.knockbackResistance", "generic.knockback_resistance")
+        .put("generic.movementSpeed", "generic.movement_speed")
+        .put("generic.flyingSpeed", "generic.flying_speed")
+        .put("generic.attackDamage", "generic.attack_damage")
+        .put("generic.attackKnockback", "generic.attack_knockback")
+        .put("generic.attackSpeed", "generic.attack_speed")
+        .put("generic.armorToughness", "generic.armor_toughness").build();
+    private static final java.util.Map<Integer, String> catTypeConversion = new com.google.common.collect.ImmutableMap.Builder<Integer, String>()
+            .put(0, "minecraft:tabby")
+            .put(1, "minecraft:black")
+            .put(2, "minecraft:red")
+            .put(3, "minecraft:siamese")
+            .put(4, "minecraft:british")
+            .put(5, "minecraft:calico")
+            .put(6, "minecraft:persian")
+            .put(7, "minecraft:ragdoll")
+            .put(8, "minecraft:white")
+            .put(9, "minecraft:jellie")
+            .put(10, "minecraft:all_black").build();
+    private static void convertNBT(@Nullable CompoundTag nbttagcompound) {
+        if (nbttagcompound != null && nbttagcompound.contains("EntityTag", 10)) {
+            CompoundTag entityTag = nbttagcompound.getCompound("EntityTag");
+            if (entityTag.contains("Attributes", 9)) {
+                ListTag nbttaglist = entityTag.getList("Attributes", 10);
+                for (int i = 0; i < nbttaglist.size(); i++) {
+                    String value = attributeConversion.get(nbttaglist.getCompound(i).getString("Name"));
+                    if (value != null) {
+                        nbttaglist.getCompound(i).putString("Name", value);
+                    }
+                }
+            }
+            if (entityTag.contains("CatType")) {
+                Integer id = entityTag.getInt("CatType");
+                String value = catTypeConversion.get(id);
+                if (value == null) {
+                    value = "minecraft:tabby";
+                }
+                entityTag.remove("CatType");
+                entityTag.putString("variant", value);
+            }
+        }
+    }
+    // EMC end
     @Nullable
     public T create(ServerLevel world, @Nullable CompoundTag itemNbt, @Nullable Component name, @Nullable Player player, BlockPos pos, MobSpawnType spawnReason, boolean alignPosition, boolean invertY) {
         T t0 = this.create(world);
